api/src/main/java/org/pmiops/workbench/google/DirectoryServiceImpl.java

```diff
@@ -102,39 +128,64 @@ public User getUser(String email) {
 
   @Override
   public boolean isUsernameTaken(String username) {
-    String gSuiteDomain = configProvider.get().googleDirectoryService.gSuiteDomain;
-    return getUser(username + "@" + gSuiteDomain) != null;
+    return getUserByUsername(username) != null;
+  }
+
+  // Returns a user's contact email address via the custom schema in the directory API.
+  public String getContactEmailAddress(String username) {
+    return (String) getUserByUsername(username)
+        .getCustomSchemas()
+        .get(GSUITE_AOU_SCHEMA_NAME)
+        .get(GSUITE_FIELD_CONTACT_EMAIL);
   }
 
   @Override
   public User createUser(String givenName, String familyName, String username, String contactEmail) {
-    String gSuiteDomain = configProvider.get().googleDirectoryService.gSuiteDomain;
+    String primaryEmail = username + "@" + gSuiteDomain();
     String password = randomString();
+
+    // GSuite custom fields for Workbench user accounts.
+    // See the Moodle integration doc (broad.io/aou-moodle) for more details, as this
+    // was primarily set up for Moodle SSO integration.
+    Map<String, Object> aouCustomFields = new HashMap<String, Object>();
+    // The value of this field must match one of the allowed values in the Moodle installation.
+    // Since this value is unlikely to ever change, we use a hard-coded constant rather than an env
+    // variable.
+    aouCustomFields.put(GSUITE_FIELD_INSTITUTION, INSTITUTION_FIELD_VALUE);
+    // This gives us a structured place to store researchers' contact email addresses, in
+    // case we want to pass it to other systems (e.g. Zendesk or Moodle) via SAML mapped fields.
+    aouCustomFields.put(GSUITE_FIELD_CONTACT_EMAIL, contactEmail);
+
     User user = new User()
-        .setPrimaryEmail(username+"@"+gSuiteDomain)
+        .setPrimaryEmail(primaryEmail)
         .setPassword(password)
         .setName(new UserName().setGivenName(givenName).setFamilyName(familyName))
-        .setEmails(new UserEmail().setType("custom").setAddress(contactEmail).setCustomType("contact"))
+        // In addition to the custom schema value, we store each user's contact email as a secondary
+        // email address with type "home". This makes it show up nicely in GSuite admin as the
+        // user's "Secondary email".
+        .setEmails(Lists.newArrayList(
+            new UserEmail().setType("work").setAddress(primaryEmail).setPrimary(true),
+            new UserEmail().setType("home").setAddress(contactEmail)))
+        .setCustomSchemas(Collections.singletonMap(GSUITE_AOU_SCHEMA_NAME, aouCustomFields))
         .setChangePasswordAtNextLogin(true);
     retryHandler.run((context) -> getGoogleDirectoryService().users().insert(user).execute());
     return user;
   }
 
   @Override
-  public User resetUserPassword(String userKey) {
-    User user = getUser(userKey);
+  public User resetUserPassword(String email) {
```

Comment:
Side note: I tried to clean up the terminology (at least within this file), so "email" is always the full email address "gjordan@researchallofus.net" and "username" is just the first part, "gjordan". In other files (and our SQL database) the distinction is not always so clear, and even Google isn't super specific about what is required for some of these API calls. For example, from the [API docs](https://developers.google.com/admin-sdk/directory/v1/guides/manage-users#update_user):

> The userKey can be the user's primary email address, the unique user id, or one of the user's alias email addresses. 

In my tests using the APIs explorer, the userKey parameter must be the full email address of the user.